### PR TITLE
Sanitize host field prior to auth flow

### PR DESCRIPTION
**Pull request by pietern**
I was getting 401s when trying to use the accounts API on GCP, until I realized the host field in the provider configuration had a trailing slash. Removing the trailing slash fixed the issue. To improve the UX here, this PR removes a trailing slash prior to the authentication flow.

### CODE REVIEW

**Review by nfx — CHANGES_REQUESTED** [review]

File: common/client.go [hunk 1]
```diff
@@ -326,12 +331,32 @@ func (c *DatabricksClient) niceAuthError(message string) error {
 	return fmt.Errorf("%s%s. Please check %s for details", message, info, docUrl)
 }
 
-func (c *DatabricksClient) fixHost() {
-	if c.Host != "" && !(strings.HasPrefix(c.Host, "https://") || strings.HasPrefix(c.Host, "http://")) {
-		// azurerm_databricks_workspace.*.workspace_url is giving URL without scheme
-		// so that is why this line is here
-		c.Host = "https://" + c.Host
+func (c *DatabricksClient) fixHost() error {
+	if c.Host != "" {
```
[the remark this answers is not in the document]
Comment (pietern) [reply common/client.go]: Done.

**Review by nfx — APPROVED** [review]

File: common/client.go [hunk 2]
```diff
@@ -271,7 +277,6 @@ func (c *DatabricksClient) Authenticate(ctx context.Context) error {
 		log.Printf("[INFO] Configured %s auth: %s", auth.name, c.configDebugString()) // lgtm[go/clear-text-logging]
 		c.authVisitor = authorizer
 		c.AuthType = auth.name
-		c.fixHost()
```
Comment (nfx) [inline common/client.go:274]: can we fix it before and after the auth resolution? e.g. misconfigured databrickscfg

**Review by nfx — CHANGES_REQUESTED** [review]

File: common/client.go [hunk 3]
```diff
@@ -271,7 +277,6 @@ func (c *DatabricksClient) Authenticate(ctx context.Context) error {
 		log.Printf("[INFO] Configured %s auth: %s", auth.name, c.configDebugString()) // lgtm[go/clear-text-logging]
 		c.authVisitor = authorizer
 		c.AuthType = auth.name
```
Comment (nfx) [inline common/client.go:279]:
```suggestion
		c.AuthType = auth.name
		c.fixHost()
```

can we fix it before and after the auth resolution? e.g. misconfigured databrickscfg

**Review by nfx — CHANGES_REQUESTED** [review]

File: common/client.go [hunk 4]
```diff
@@ -326,12 +331,34 @@ func (c *DatabricksClient) niceAuthError(message string) error {
...
+	if err != nil {
+		return err
+	}
+
+	// If the host is empty, assume the scheme wasn't included.
+	if u.Host == "" {
+		u, err = url.Parse("https://" + c.Host)
+		if err != nil {
+			return err
+		}
```
Comment (nfx) [inline common/client.go:350]: this is strange, as invocation will always error out because of `url.Parse("https://")`

File: common/client.go [hunk 5]
```diff
@@ -326,12 +331,34 @@ func (c *DatabricksClient) niceAuthError(message string) error {
...
 	}
+
+	// Create new instance to ensure other fields are initialized as empty.
+	u = &url.URL{
+		Scheme: u.Scheme,
+		Host:   u.Host,
+	}
+
+	// Store sanitized version of c.Host.
+	c.Host = u.String()
```
Comment (nfx) [inline common/client.go:360]: why not just `fmt.Sprintf("https://%s", c.Host)`?

File: common/client.go (hunk 4 above)
Comment (pietern) [reply common/client.go:350]: No, u is the parsed URL, c is the input URL.

Comment (nfx) [reply common/client.go:351]: ah, it's `u`, i've read it as `c`.

File: common/client.go (hunk 5 above)
Comment (nfx) [reply common/client.go:361]: `TestConfig_PatFromDatabricksCfg` is failing. as long as we're doing https://github.com/databricks/terraform-provider-databricks/blob/master/common/http.go#L338-L343, it's safe to update the test

**Pull request merged**